crates/subspace-core-primitives/src/lib.rs

```diff
@@ -95,23 +93,122 @@ pub type Blake3Hash = [u8; BLAKE3_HASH_SIZE];
     Hash,
     From,
     Into,
+    AsRef,
+    AsMut,
     Deref,
     DerefMut,
     Encode,
     Decode,
     TypeInfo,
     MaxEncodedLen,
 )]
+#[into(ref, ref_mut)]
+#[as_ref(forward)]
+#[as_mut(forward)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(transparent))]
-pub struct Blake3HashHex(#[cfg_attr(feature = "serde", serde(with = "hex"))] Blake3Hash);
+pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; BLAKE3_HASH_SIZE]);
 
-impl fmt::Debug for Blake3HashHex {
+impl fmt::Debug for Blake3Hash {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{}", hex::encode(self.0))
     }
 }
 
+impl FromHex for Blake3Hash {
+    type Error = hex::FromHexError;
+
+    fn from_hex<T: AsRef<[u8]>>(hex: T) -> Result<Self, Self::Error> {
+        let data = hex::decode(hex)?
+            .try_into()
+            .map_err(|_| hex::FromHexError::InvalidStringLength)?;
+
+        Ok(Self(data))
+    }
+}
+
+impl From<Blake3Hash> for Vec<u8> {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        hash.0.to_vec()
+    }
+}
+
+impl From<Blake3Hash> for PosSeed {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        Self(hash.0)
+    }
+}
+
+impl From<Blake3Hash> for Randomness {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        Self::from(hash.0)
+    }
+}
+
+impl From<&[u8; BLAKE3_HASH_SIZE]> for Blake3Hash {
+    #[inline]
+    fn from(value: &[u8; BLAKE3_HASH_SIZE]) -> Self {
+        Self(*value)
+    }
+}
+
+impl TryFrom<&[u8]> for Blake3Hash {
+    type Error = TryFromSliceError;
+
+    #[inline]
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        Ok(Self(value.try_into()?))
+    }
+}
+
+impl TryFrom<Vec<u8>> for Blake3Hash {
+    type Error = TryFromSliceError;
+
+    #[inline]
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        value.as_slice().try_into()
+    }
+}
+
+impl PartialEq<&[u8]> for Blake3Hash {
+    #[inline]
+    fn eq(&self, other: &&[u8]) -> bool {
+        self.0 == *other
+    }
+}
+
+impl PartialEq<Blake3Hash> for &[u8] {
+    #[inline]
+    fn eq(&self, other: &Blake3Hash) -> bool {
+        *self == other.0
+    }
+}
```

Comment:
I don't think we need any of these:
```suggestion
```

Reply:
I think the `TryFrom` is required because of the incorrect merkle tree associated type.

The `PartialEq`s can be done using dereferencing, if we're happy for the API to be less ergonomic than `[u8; 32]`'s.

Reply:
Then we should fix merkle tree (which doesn't actually use Blake3) and for comparison we really should compare two values of identical types. Comparing slice of bytes to Blake3Hash while convenient at times is a slightly confusing API and I'd argue one should dereference, use `.as_ref()` or something like that instead of requiring custom impls like this.